base/utils/vmaas.go

```diff
@@ -0,0 +1,158 @@
+	vmaasDataB *vmaas.UpdatesV2Response) (*vmaas.UpdatesV2Response, error) {
+	if *vmaasDataA.Basearch != *vmaasDataB.Basearch {
+		return nil, errors.New("unable to merge different archs")
+	}
+
+	if *vmaasDataA.Releasever != *vmaasDataB.Releasever {
+		return nil, errors.New("release versions do not match")
+	}
+
+	if !RepoListsEqual(vmaasDataA.RepositoryList, vmaasDataB.RepositoryList) {
+		return nil, errors.New("repository lists differ")
+	}
+	if !ModuleListsEqual(vmaasDataA.ModulesList, vmaasDataB.ModulesList) {
+		return nil, errors.New("module lists differ")
+	}
+
+	mergedList := vmaasDataA.GetUpdateList()
+	for nevraB, updateListB := range vmaasDataB.GetUpdateList() {
+		if updateListA, ok := mergedList[nevraB]; ok {
+			merged, err := mergeUpdates(updateListA, updateListB)
+			if err != nil {
+				return nil, err
+			}
+			mergedList[nevraB] = *merged
+		} else {
+			mergedList[nevraB] = updateListB
+		}
+	}
+
+	vmaasDataA.UpdateList = &mergedList
+
+	return vmaasDataA, nil
+}
+
+func mergeUpdates(listA, listB vmaas.UpdatesV2ResponseUpdateList) (*vmaas.UpdatesV2ResponseUpdateList, error) {
+	updatesA, updatesB := listA.GetAvailableUpdates(), listB.GetAvailableUpdates()
+	if len(updatesA) == 0 && len(updatesB) == 0 {
+		return nil, nil
+	}
+	if len(updatesA) == 0 {
+		return &listB, nil
+	}
+	if len(updatesB) == 0 {
+		return &listA, nil
+	}
+
+	newUpdates, err := processMerge(updatesA, updatesB)
+	if err != nil {
+		return nil, err
+	}
+
+	return &vmaas.UpdatesV2ResponseUpdateList{
+		AvailableUpdates: newUpdates,
+	}, nil
+}
+
+func processMerge(updatesA,
+	updatesB []vmaas.UpdatesV2ResponseAvailableUpdates) (*[]vmaas.UpdatesV2ResponseAvailableUpdates, error) {
+	newUpdates := make([]vmaas.UpdatesV2ResponseAvailableUpdates, 0)
+	processed := make(map[string]bool)
+	offset, i := 0, 0
+	for i < len(updatesA) {
+		nevraA, err := ParseNevra(updatesA[i].GetPackage())
+		if err != nil {
+			return nil, err
+		}
+		if offset > len(updatesB)-1 {
+			appendUpdateIfNotExists(&newUpdates, updatesA[i], nevraA.Version, processed)
+		}
+		for j := 0 + offset; j < len(updatesB); j++ {
+			nevraB, err := ParseNevra(updatesB[j].GetPackage())
+			if err != nil {
+				return nil, err
+			}
+			cmp := nevraA.Cmp(nevraB)
+
+			if cmp < 0 {
+				appendUpdateIfNotExists(&newUpdates, updatesA[i], nevraA.Version, processed)
+				break
+			}
+			if cmp > 0 {
+				appendUpdateIfNotExists(&newUpdates, updatesB[j], nevraB.Version, processed)
+				offset++
+				if offset == len(updatesB) {
+					appendUpdateIfNotExists(&newUpdates, updatesA[i], nevraA.Version, processed)
+				}
+				continue
+			}
+			appendUpdateIfNotExists(&newUpdates, updatesA[i], nevraA.Version, processed)
+			appendUpdateIfNotExists(&newUpdates, updatesB[j], nevraB.Version, processed)
+			offset++
+			break
+		}
+		i++
+	}
+	if offset <= len(updatesB)-1 {
+		newUpdates = append(newUpdates, updatesB[offset:]...)
+	}
+	return &newUpdates, nil
+}
```

Comment:
The merge loop works (according to the tests) but it's unnecessary hard to understand.
I've rewritten it into something cleaner (hopefully).

Reply:
Actually there's a bug in it, see my comment in the kernel test.

Reply:
Yes, that function was simple before speed optimization, definitely not after.

Reply:
Very nice, much cleaner!